exec/java-exec/src/main/java/org/apache/drill/exec/ssl/SSLCredentialsProvider.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.ssl;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+/**
+ * Provides an implementation of credentials provider depending on "useMapRSSLConfig" option value
+ * and whether Drill was built with mapr profile or not.
+ */
+abstract class SSLCredentialsProvider {
+
+  private static final String MAPR_CREDENTIALS_PROVIDER_CLIENT = "org.apache.drill.exec.ssl.SSLCredentialsProviderMaprClient";
+  private static final String MAPR_CREDENTIALS_PROVIDER_SERVER = "org.apache.drill.exec.ssl.SSLCredentialsProviderMaprServer";
+
+  /**
+   * Provides a concrete implementation of {@link SSLCredentialsProvider}.
+   *
+   * @param getPropertyMethod a reference to a method to retrieve credentials from config:
+   *                          <ul>
+   *                            <li>String parameter1 - property name</li>
+   *                            <li>String parameter2 - default value</li>
+   *                            <li>returns the property value or default value</li>
+   *                          </ul>
+   * @param mode              CLIENT or SERVER
+   * @param useMapRSSLConfig  use a MapR credential provider
+   * @return concrete implementation of SSLCredentialsProvider
+   */
+  static SSLCredentialsProvider getSSLCredentialsProvider(BiFunction<String, String, String> getPropertyMethod, SSLConfig.Mode mode, boolean useMapRSSLConfig) {
+    return useMapRSSLConfig ?
+        Optional.ofNullable(getMaprCredentialsProvider(mode))
```

Comment:
It would be better to return optional from ```getMaprCredentialsProvider(mode)``` and use ```.orElseGet(supplier)``` to avoid creating everytime instance of ```SSLCredentialsProviderImpl```.